larq_compute_engine/tflite/python/interpreter_wrapper_utils.h

```diff
@@ -0,0 +1,201 @@
+pybind11::dtype PyTypeFromTfLiteType(TfLiteType type) {
+  switch (type) {
+    case kTfLiteFloat32:
+      return pybind11::dtype::of<float>();
+    case kTfLiteUInt8:
+      return pybind11::dtype::of<std::uint8_t>();
+    case kTfLiteInt8:
+      return pybind11::dtype::of<std::int8_t>();
+    case kTfLiteInt16:
+      return pybind11::dtype::of<std::int16_t>();
+    case kTfLiteInt32:
+      return pybind11::dtype::of<std::int32_t>();
+    case kTfLiteInt64:
+      return pybind11::dtype::of<std::int64_t>();
+    case kTfLiteBool:
+      return pybind11::dtype::of<bool>();
+    case kTfLiteString:
+      return pybind11::dtype::of<char*>();
+    case kTfLiteNoType:
+    default:
+      PY_ERROR("Model has invalid output type: " << type);
+      return pybind11::dtype::of<float>();
+  };
+}
+
+bool SetTensorFromNumpy(const TfLiteTensor* tensor,
+                        const pybind11::array& nparray) {
+  TfLiteType type = TfLiteTypeFromPyType(nparray.dtype());
+
+  if (type != tensor->type) {
+    PY_ERROR("Expected tensor type " << TfLiteTypeGetName(tensor->type)
+                                     << " but got " << TfLiteTypeGetName(type)
+                                     << " for tensor " << tensor->name);
+    return false;
+  }
+
+  int ndim = nparray.ndim();
+  if (ndim != tensor->dims->size) {
+    PY_ERROR("Expected tensor dimension " << tensor->dims->size << " but got "
+                                          << ndim << " for tensor "
+                                          << tensor->name);
+    return false;
+  }
+
+  for (int j = 0; j < ndim; j++) {
+    if (tensor->dims->data[j] != nparray.shape(j)) {
+      PY_ERROR("Expected " << tensor->dims->data[j] << " for dimension " << j
+                           << " but found " << nparray.shape(j)
+                           << " for tensor " << tensor->name);
+      return false;
+    }
+  }
+
+  size_t size = nparray.nbytes();
+  if (size != tensor->bytes) {
+    PY_ERROR("Expected " << tensor->bytes << " bytes but got " << size
+                         << " for tensor " << tensor->name);
+    return false;
+  }
+  memcpy(tensor->data.raw, nparray.data(), size);
+  return true;
+}
+
+template <typename InterpreterType>
+pybind11::list InterpreterWrapperBase<InterpreterType>::get_types(
+    const std::vector<int>& tensors) const {
+  pybind11::list result;
+
+  for (auto tensor_id : tensors) {
+    const TfLiteTensor* tensor = interpreter_->tensor(tensor_id);
+    result.append(PyTypeFromTfLiteType(tensor->type));
+  }
+
+  return result;
+}
+
+template <typename InterpreterType>
+pybind11::list InterpreterWrapperBase<InterpreterType>::get_shapes(
+    const std::vector<int>& tensors) const {
+  pybind11::list result;
+
+  for (auto tensor_id : tensors) {
+    const TfLiteTensor* tensor = interpreter_->tensor(tensor_id);
+    pybind11::tuple shape(tensor->dims->size - 1);
+    for (int j = 0; j < tensor->dims->size - 1; ++j)
+      shape[j] = tensor->dims->data[j + 1];
```

Comment:
This won't return the empty batch dimension now, but I'd be open to revert this to the previous behaviour.

Reply:
I'd rather leave this the way it was and have the C++ pybind be as generic as possible, and then 'fix' it on the python side. TFLite has full support for `batch != 1`, and with this change we'd not support that.

Reply:
> TFLite has full support for `batch != 1`, and with this change we'd not support that.

Ah didn't know that since I only ever saw models with `batch == 1`.  Should we keep the Python API then the same as well and always return `[batch, *tensor_shape]` which we get from TFLite?

Reply:
That is fine with me, I don't feel too strongly about that.

Reply:
OK, I reverted it to the previous behaviour. I think this now also roughly matches what Keras does:
```python
In [9]: input_shape = (24, 24, 3)
   ...: x = tf.keras.Input(input_shape)
   ...: model = tf.keras.Model(x, tf.keras.layers.Flatten()(x))

In [10]: model.input_shape
Out[10]: (None, 24, 24, 3)

In [11]: model.output_shape
Out[11]: (None, 1728)
```